c10/core/UndefinedTensorImpl.h

```diff
@@ -17,14 +17,7 @@ struct C10_API UndefinedTensorImpl final : public TensorImpl {
 #endif
     return &_singleton;
   }
-  IntArrayRef sizes() const override;
-  IntArrayRef strides() const override;
-  int64_t size(int64_t d) const override;
-  int64_t stride(int64_t d) const override;
-  int64_t dim() const override;
   bool has_storage() const override;
```

Comment:
why is has_storage still virtual?

Reply:
I am thinking whether the derived types of TensorImpl is responsible for telling whether it has storage. An alternative is to explicitly check whether the current tensor is sparse or MKLDNN tensor, and if so return false in `has_storage()`.

Reply:
I changed it to explicitly check whether the current tensor is sparse or MKLDNN tensor, and if so return false in `has_storage()`.